src/test/java/jenkins/scm/impl/PersistAcrossRestartsTest.java

```diff
@@ -0,0 +1,176 @@
+package jenkins.scm.impl;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.model.FreeStyleProject;
+import hudson.model.ItemGroup;
+import hudson.model.Job;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.TopLevelItem;
+import hudson.model.TopLevelItemDescriptor;
+import hudson.scm.NullSCM;
+import hudson.scm.SCM;
+import jenkins.model.Jenkins;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMHeadEvent;
+import jenkins.scm.api.SCMHeadObserver;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.SCMSourceCriteria;
+import jenkins.scm.api.SCMSourceOwner;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runners.model.Statement;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+import org.jvnet.hudson.test.TestExtension;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+
+public class PersistAcrossRestartsTest {
+    @Rule
+    public RestartableJenkinsRule story = new RestartableJenkinsRule();
+
+    @Issue("JENKINS-48571")
+    @Test
+    public void uuidIdPersists() throws Exception {
+        story.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                BogusSCMSource source = new BogusSCMSource();
+                SCMItem p = story.j.jenkins.createProject(SCMItem.class, "test-with-source");
+                p.setSource(source);
+                // TODO: Find a better way to persist this for testing across restart.
+                FreeStyleProject p2 = story.j.createFreeStyleProject("test-without-source");
```

Comment:
Just stash it in an AtomicReference in the Test class. The test class instance is durable for the duration of the test